File: src/handlers/commands.py
"""Slash command handlers for WeeklyRoulette bot."""

import asyncio

from slack_bolt import Ack, App
from slack_bolt.context.respond import Respond
from slack_sdk import WebClient

from database.models import DAY_DISPLAY_NAMES, VALID_DAYS
from services.roulette import RouletteService
from services.scheduler import SchedulerService


def register_command_handlers(
    app: App, roulette_service: RouletteService, scheduler_service: SchedulerService
) -> None:
    """Register all slash command handlers."""

    @app.command("/weeklyroulette")
    def handle_weeklyroulette_command(
        ack: Ack, command: dict, client: WebClient, respond: Respond
    ) -> None:
        """Handle /weeklyroulette slash command."""
        ack()

        text = command.get("text", "").strip()
        channel_id = command["channel_id"]

        if not text or text == "help":
            help_message = (
                "🎲 **WeeklyRoulette - Help**\n\n"
                "**Available commands:**\n"
                "• `/weeklyroulette config` - configure selection schedule\n"
                "• `/weeklyroulette status` - check current settings\n"
                "• `/weeklyroulette help` - show this help\n\n"
                "**How it works:**\n"
                "1. Configure day and time for selection\n"
                "2. Bot automatically picks a random person from channel\n"
                "3. Selection happens every week at the set time\n\n"
            )
            respond(text=help_message)
            return

        if text == "config":
            _show_config_modal(client, command["trigger_id"], channel_id)
            return

        if text == "test":
            result = asyncio.run(
                roulette_service.send_roulette_message(channel_id, test_mode=True)
            )

            if result["success"]:
                respond(
                    text=f"✅ Test completed successfully!\n\n{result['message']}\n\n"
                    f"💡 *This was just a test. Regular selection will happen automatically according to schedule.*"
                )
            else:
                respond(text=f"❌ Test failed: {result['error']}")
            return

        if text == "status":
            status = roulette_service.get_channel_status(channel_id)
            respond(text=status["message"])
            return

        respond(
            text="❓ Unknown command. Use `/weeklyroulette help` to see available options."
        )

    def _show_config_modal(client: WebClient, trigger_id: str, channel_id: str) -> None:
        """Show configuration modal dialog."""

        current_config = roulette_service.db.get_channel_config(channel_id)

        day_options = []
        for day in VALID_DAYS:
            day_options.append(
                {
                    "text": {"type": "plain_text", "text": DAY_DISPLAY_NAMES[day]},
                    "value": day,
                }
            )

        time_options = []
        for hour in range(9, 17):  # 9:00 to 16:55 (96 options, under Slack's 100 limit)
            for minute in range(0, 60, 5):  # 5-minute intervals
                time_str = f"{hour:02d}:{minute:02d}"
                display_time = f"{hour:02d}:{minute:02d}"
                time_options.append(
                    {
                        "text": {"type": "plain_text", "text": display_time},
                        "value": time_str,
                    }
                )

        default_day = current_config.day if current_config else "friday"
        default_time = current_config.time if current_config else "15:00"
        default_enabled = current_config.enabled if current_config else True

        modal_view = {
            "type": "modal",
            "callback_id": "config_modal",
            "title": {"type": "plain_text", "text": "Configuration"},
            "submit": {"type": "plain_text", "text": "Save"},
            "close": {"type": "plain_text", "text": "Cancel"},
            "private_metadata": channel_id,
            "blocks": [
                {
                    "type": "section",
                    "text": {
                        "type": "mrkdwn",
                        "text": "🎲 *Configure weekly selection schedule*",
                    },
                },
                {"type": "divider"},
                {
                    "type": "section",
                    "text": {"type": "mrkdwn", "text": "📅 *Choose day of week:*"},
                    "accessory": {
                        "type": "static_select",
                        "placeholder": {"type": "plain_text", "text": "Select day"},
                        "action_id": "day_select",
                        "initial_option": {
                            "text": {
                                "type": "plain_text",
                                "text": DAY_DISPLAY_NAMES[default_day],
                            },
                            "value": default_day,
                        },
                        "options": day_options,
                    },
                },
                {
                    "type": "section",
                    "text": {
                        "type": "mrkdwn",
                        "text": "🕒 *Choose time (Polish time):*",
                    },
                    "accessory": {
                        "type": "static_select",
                        "placeholder": {
                            "type": "plain_text",
                            "text": "Select time",
                        },
                        "action_id": "time_select",
                        "initial_option": {
                            "text": {"type": "plain_text", "text": default_time},
                            "value": default_time,
                        },
                        "options": time_options,
                    },
                },
                {
                    "type": "section",
                    "text": {"type": "mrkdwn", "text": "🔄 *Status:*"},
                    "accessory": {
                        "type": "static_select",
                        "placeholder": {"type": "plain_text", "text": "Select status"},
                        "action_id": "enabled_select",
                        "initial_option": {
                            "text": {
                                "type": "plain_text",
                                "text": "Enabled" if default_enabled else "Disabled",
                            },
                            "value": "true" if default_enabled else "false",
                        },
                        "options": [
                            {
                                "text": {"type": "plain_text", "text": "Enabled"},
                                "value": "true",
                            },
                            {
                                "text": {"type": "plain_text", "text": "Disabled"},
                                "value": "false",
                            },
                        ],
                    },
                },
                {"type": "divider"},
                {
                    "type": "context",
                    "elements": [
                        {
                            "type": "mrkdwn",
                            "text": "💡 *Bot will automatically select one person from the channel at the chosen time*",
                        }
                    ],
                },
            ],
        }

        try:
            client.views_open(trigger_id=trigger_id, view=modal_view)
        except Exception as e:
            print(f"Error opening config modal: {e}")

    print("✅ Command handlers registered")

File: internal/handlers/commands.go
package handlers

import (
	"context"
	"fmt"
	"log/slog"

	"github.com/slack-go/slack"
	"github.com/typeWolffo/weeklyroulette-bot/internal/database"
	"github.com/typeWolffo/weeklyroulette-bot/internal/services"
)

type CommandHandler struct {
	roulette  *services.RouletteService
	scheduler *services.SchedulerService
}

func NewCommandHandler(roulette *services.RouletteService, scheduler *services.SchedulerService) *CommandHandler {
	return &CommandHandler{
		roulette:  roulette,
		scheduler: scheduler,
	}
}

func (h *CommandHandler) Handle(cmd slack.SlashCommand, client *slack.Client) (string, error) {
	text := cmd.Text
	channelID := cmd.ChannelID

	slog.Info("received command", "command", text, "channel", channelID, "user", cmd.UserID)

	switch text {
	case "", "help":
		return h.handleHelp(), nil
	case "config":
		return h.handleConfig(cmd, client)
	case "test":
		return h.handleTest(channelID)
	case "status":
		return h.handleStatus(channelID)
	default:
		slog.Warn("unknown command", "text", text)
		return "❓ Unknown command. Use `/weeklyroulette help` to see available options.", nil
	}
}

func (h *CommandHandler) handleHelp() string {
	return `WeeklyRoulette - Help

Available commands:
• ` + "`/weeklyroulette config`" + ` - configure selection schedule
• ` + "`/weeklyroulette status`" + ` - check current settings
• ` + "`/weeklyroulette test`" + ` - run a test selection
• ` + "`/weeklyroulette help`" + ` - show this help

How it works:
1. Configure day and time for selection
2. Bot automatically picks a random person from channel
3. Selection happens every week at the set time
`
}

func (h *CommandHandler) handleConfig(cmd slack.SlashCommand, client *slack.Client) (string, error) {
	ctx := context.Background()
	currentConfig, _ := h.roulette.DB().GetChannelConfig(ctx, cmd.ChannelID)

	modal := buildConfigModal(cmd.ChannelID, currentConfig)

	_, err := client.OpenView(cmd.TriggerID, modal)
	if err != nil {
		return "", fmt.Errorf("opening config modal: %w", err)
	}

	return "", nil
}

func (h *CommandHandler) handleTest(channelID string) (string, error) {
	slog.Info("starting test roulette", "channel", channelID)

	ctx := context.Background()
	result, err := h.roulette.SendRouletteMessage(ctx, channelID, true)
	if err != nil {
		slog.Error("test roulette failed with error", "channel", channelID, "error", err)
		return "", err
	}

	if result.Success {
		slog.Info("test roulette completed",
			"channel", channelID,
			"selected_user", result.SelectedMember.Name,
			"selected_id", result.SelectedMember.ID,
			"total_members", result.TotalMembers,
		)
		return fmt.Sprintf("✅ Test completed successfully!\n\n%s\n\n💡 *This was just a test. Regular selection will happen automatically according to schedule.*", result.Message), nil
	}

	slog.Warn("test roulette failed", "channel", channelID, "error", result.Error)
	return fmt.Sprintf("❌ Test failed: %s", result.Error), nil
}

func (h *CommandHandler) handleStatus(channelID string) (string, error) {
	ctx := context.Background()
	status := h.roulette.GetChannelStatus(ctx, channelID)
	return status["message"].(string), nil
}

func buildConfigModal(channelID string, currentConfig *database.ChannelConfig) slack.ModalViewRequest {
	defaultDay := "friday"
	defaultTime := "15:00"
	defaultEnabled := true

	if currentConfig != nil {
		defaultDay = currentConfig.Day
		defaultTime = currentConfig.Time
		defaultEnabled = currentConfig.Enabled
	}

	dayOptions := make([]*slack.OptionBlockObject, 0, len(database.ValidDays))
	for _, day := range database.ValidDays {
		displayName := database.DayDisplayNames[day]
		dayOptions = append(dayOptions, slack.NewOptionBlockObject(
			day,
			slack.NewTextBlockObject(slack.PlainTextType, displayName, false, false),
			nil,
		))
	}

	var timeOptions []*slack.OptionBlockObject
	for hour := 9; hour < 17; hour++ {
		for minute := 0; minute < 60; minute += 5 {
			timeStr := fmt.Sprintf("%02d:%02d", hour, minute)
			timeOptions = append(timeOptions, slack.NewOptionBlockObject(
				timeStr,
				slack.NewTextBlockObject(slack.PlainTextType, timeStr, false, false),
				nil,
			))
		}
	}

	var initialDayOption *slack.OptionBlockObject
	for _, opt := range dayOptions {
		if opt.Value == defaultDay {
			initialDayOption = opt
			break
		}
	}

	var initialTimeOption *slack.OptionBlockObject
	for _, opt := range timeOptions {
		if opt.Value == defaultTime {
			initialTimeOption = opt
			break
		}
	}

	enabledText := "Enabled"
	enabledValue := "true"

	if !defaultEnabled {
		enabledText = "Disabled"
		enabledValue = "false"
	}

	initialEnabledOption := slack.NewOptionBlockObject(
		enabledValue,
		slack.NewTextBlockObject(slack.PlainTextType, enabledText, false, false),
		nil,
	)

	blocks := slack.Blocks{
		BlockSet: []slack.Block{
			slack.NewSectionBlock(
				slack.NewTextBlockObject(slack.MarkdownType, "🎲 *Configure weekly selection schedule*", false, false),
				nil, nil,
			),
			slack.NewDividerBlock(),
			slack.NewSectionBlock(
				slack.NewTextBlockObject(slack.MarkdownType, "📅 *Choose day of week:*", false, false),
				nil,
				slack.NewAccessory(slack.NewOptionsSelectBlockElement(
					slack.OptTypeStatic,
					slack.NewTextBlockObject(slack.PlainTextType, "Select day", false, false),
					"day_select",
					dayOptions...,
				).WithInitialOption(initialDayOption)),
			),
			slack.NewSectionBlock(
				slack.NewTextBlockObject(slack.MarkdownType, "🕒 *Choose time (Polish time):*", false, false),
				nil,
				slack.NewAccessory(slack.NewOptionsSelectBlockElement(
					slack.OptTypeStatic,
					slack.NewTextBlockObject(slack.PlainTextType, "Select time", false, false),
					"time_select",
					timeOptions...,
				).WithInitialOption(initialTimeOption)),
			),
			slack.NewSectionBlock(
				slack.NewTextBlockObject(slack.MarkdownType, "🔄 *Status:*", false, false),
				nil,
				slack.NewAccessory(slack.NewOptionsSelectBlockElement(
					slack.OptTypeStatic,
					slack.NewTextBlockObject(slack.PlainTextType, "Select status", false, false),
					"enabled_select",
					slack.NewOptionBlockObject("true", slack.NewTextBlockObject(slack.PlainTextType, "Enabled", false, false), nil),
					slack.NewOptionBlockObject("false", slack.NewTextBlockObject(slack.PlainTextType, "Disabled", false, false), nil),
				).WithInitialOption(initialEnabledOption)),
			),
			slack.NewDividerBlock(),
			slack.NewContextBlock("",
				slack.NewTextBlockObject(slack.MarkdownType, "💡 *Bot will automatically select one person from the channel at the chosen time*", false, false),
			),
		},
	}

	return slack.ModalViewRequest{
		Type:            slack.VTModal,
		CallbackID:      "config_modal",
		Title:           slack.NewTextBlockObject(slack.PlainTextType, "Configuration", false, false),
		Submit:          slack.NewTextBlockObject(slack.PlainTextType, "Save", false, false),
		Close:           slack.NewTextBlockObject(slack.PlainTextType, "Cancel", false, false),
		PrivateMetadata: channelID,
		Blocks:          blocks,
	}
}
